api/paidAction/index.js

```diff
@@ -107,283 +202,285 @@ export default async function performPaidAction (actionType, args, context) {
-  }
-
-  // just create the invoice and complete action when it's paid
-  const invoiceArgs = context.invoiceArgs ?? await createSNInvoice(actionType, args, context)
-  return {
-    invoice: await createDbInvoice(actionType, args, context, invoiceArgs),
-    paymentMethod: 'PESSIMISTIC'
   }
+  if (context.tx) return await performInvoicedAction(context.tx)
+  return await models.$transaction(performInvoicedAction, { isolationLevel: Prisma.TransactionIsolationLevel.Serializable })
 }
 
-async function performP2PAction (actionType, args, context) {
-  const { me } = context
-  const invoiceArgs = await createWrappedInvoice(actionType, args, context)
-  context.invoiceArgs = invoiceArgs
-
-  return me
-    ? await performOptimisticAction(actionType, args, context)
-    : await performPessimisticAction(actionType, args, context)
-}
-
-export async function retryPaidAction (actionType, args, context) {
-  const { models, me } = context
-  const { invoice: failedInvoice } = args
-
-  console.log('retryPaidAction', actionType, args)
-
-  const action = paidActions[actionType]
-  if (!action) {
-    throw new Error(`retryPaidAction - invalid action type ${actionType}`)
-  }
-
-  if (!me) {
-    throw new Error(`retryPaidAction - must be logged in ${actionType}`)
-  }
-
-  if (!action.paymentMethods.includes(PAID_ACTION_PAYMENT_METHODS.OPTIMISTIC)) {
-    throw new Error(`retryPaidAction - action does not support optimism ${actionType}`)
-  }
-
-  if (!action.retry) {
-    throw new Error(`retryPaidAction - action does not support retrying ${actionType}`)
-  }
-
-  if (!failedInvoice) {
-    throw new Error(`retryPaidAction - missing invoice ${actionType}`)
+async function performPessimisticAction (invoiceData, actionType, paidAction, args, context) {
+  const { models, actionId, me } = context
+  const performInvoicedAction = async tx => {
+    const subContext = { ...context, tx, optimistic: false }
+    await checkUser(me?.id ?? USER_ID.anon, subContext)
+    const invoiceEntry = await createDbInvoice(invoiceData, {
+      actionType,
+      actionId,
+      optimistic: false,
+      args
+    }, subContext)
+    return {
+      invoice: invoiceEntry,
+      paymentMethod: 'PESSIMISTIC',
+      retriableWithWallet: invoiceData.retriable
+    }
   }
+  if (context.tx) return await performInvoicedAction(context.tx)
+  return await models.$transaction(performInvoicedAction, { isolationLevel: Prisma.TransactionIsolationLevel.Serializable })
+}
 
-  context.optimistic = true
-  context.me = await models.user.findUnique({ where: { id: me.id } })
-
-  const { msatsRequested, actionId } = failedInvoice
-  context.cost = BigInt(msatsRequested)
-  context.actionId = actionId
-  const invoiceArgs = await createSNInvoice(actionType, args, context)
+export async function retryPaidAction ({ invoiceId, forceInternal, resetAttempts, prioritizeInternal }, context) {
+  const { models, lnd, boss, me } = context
 
   return await models.$transaction(async tx => {
```

Comment:
I am not sure if running everything inside the transaction is overkill, but i suppose we don't want a race condition where the invoice is cancelled while the action is still being retried.